lib/web/apiserver.go

```diff
@@ -112,15 +113,53 @@ type Config struct {
 	// FIPS mode means Teleport started in a FedRAMP/FIPS 140-2 compliant
 	// configuration.
 	FIPS bool
+
+	// AccessPoint holds a cache to the Auth Server.
+	AccessPoint auth.AccessPoint
+
+	// Emitter is event emitter
+	Emitter events.StreamEmitter
+
+	// HostUUID is the UUID of this process.
+	HostUUID string
+
+	// Context is used to signal process exit.
+	Context context.Context
 }
 
 type RewritingHandler struct {
 	http.Handler
 	handler *Handler
+
+	// appHandler is a http.Handler to forward requests to applications.
+	appHandler *app.Handler
+
+	// publicAddr is the public address the proxy is running at.
+	publicAddr string
 }
 
-func (r *RewritingHandler) Close() error {
-	return r.handler.Close()
+// Check if this request should be forwarded to an application handler to
+// handled by the UI and handle the request appropriately.
+func (h *RewritingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// If the request is either to the fragment authentication endpoint or if the
+	// request is already authenticated (has a session cookie), forward to
+	// application handlers. If the request is unauthenticated and requesting a
+	// FQDN that is not of the proxy, redirect to application launcher.
+	if app.HasFragment(r) || app.HasSession(r) {
```

Comment:
`if app.HasFragment(r) || app.HasSession(r) `

What's the lifetime of the app session cookie? Is it possible that there is no valid session for existing cookie and if so, how do we delete the cookie then? Just want to avoid a situation when a user sees access denied error without knowing what to do (no redirects to the launcher or no explicit "logout/login" links)

Reply:
I fixed this to redirect to login instead of showing forbidden.